### PR TITLE
Fix sample selection with map panel

### PR DESCRIPTION
**Pull request by benjaminpkane**
## What changes are proposed in this pull request?

Sample ID to grid index map should be cleared when the grid is reset

## How is this patch tested? If it is not, please explain why.

Hook test ✨ 

## Release Notes

* Fixed sample selection when interacting with plugins that use an extended selection

### What areas of FiftyOne does this PR affect?

-   [x] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [ ] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [ ] Other

## Summary by CodeRabbit

- **New Features**
  - Introduced a custom hook `useRecords` for enhanced record management.
  - Added unit tests for the `useRecords` hook to ensure reliability.
  - Introduced unit tests for `addRange` and `removeRange` functions in the `useSelectSample` module.

- **Improvements**
  - Streamlined import statements for better clarity and type safety.
  - Enhanced error handling in record retrieval logic.
  - Improved flexibility and efficiency in the `useSpotlightPager` hook by adjusting its parameter structure.

### CODE REVIEW

File: app/packages/core/src/components/Grid/useRecords.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+export type Records = MutableRefObject<Map<string, number>>;
+
+export default (clear: string) => {
+  const records = useRef(new Map<string, number>());
+
+  useMemo(() => {
+    clear;
+    records.current.clear();
+  }, [clear]);
```
Comment (sashankaryal) [inline app/packages/core/src/components/Grid/useRecords.ts:12]: What's going on here? I don't get why it's a memo and not an effect.

**Review by sashankaryal — CHANGES_REQUESTED** [review]

File: app/packages/core/src/components/Grid/useRecords.ts (hunk 1 above)
Comment (benjaminpkane) [reply app/packages/core/src/components/Grid/useRecords.ts:12]: The spotlight instance is a memo, but I suppose an unmount in an effect makes more sense here 👍

Comment (benjaminpkane) [reply app/packages/core/src/components/Grid/useRecords.ts]: I think `useMemo` is the right call here. I removed the `useRef` weirdness. Now a new records `Map` is created with each new Spotlight instance. They have the same lifecycle, so I think this makes sense

**Comment by benjaminpkane** [comment]
I think the real issue popped up during cleanup. Range selection needed to omit selections that the grid doesn't know about. Added unit tests to assert behavior

File: app/packages/core/src/components/Grid/useRecords.ts (hunk 1 above)
Comment (sashankaryal) [reply app/packages/core/src/components/Grid/useRecords.ts]: `return useMemo(() => new Map(), [clear])` makes sense to me. Just that previously it was being used like an effect and I couldn't wrap my head around it 😅

**Review by sashankaryal — APPROVED** [review]
LGTM! Tests look good, too.

**Pull request merged**